metaflow/parameters.py

```diff
@@ -246,21 +246,51 @@ class MyFlow(FlowSpec):
         indicate that the value must be a valid JSON object. A function
         implies that the parameter corresponds to a *deploy-time parameter*.
         The type of the default value is used as the parameter `type`.
-    type : type
+    type : Type, default: None
         If `default` is not specified, define the parameter type. Specify
-        one of `str`, `float`, `int`, `bool`, or `JSONType` (default: str).
-    help : str
+        one of `str`, `float`, `int`, `bool`, or `JSONType`. If None, defaults to str.
+    help : str, optional
         Help text to show in `run --help`.
-    required : bool
+    required : bool, default: False
         Require that the user specified a value for the parameter.
         `required=True` implies that the `default` is not used.
-    show_default : bool
+    show_default : bool, default: True
         If True, show the default value in the help text (default: True).
     """
 
-    def __init__(self, name, **kwargs):
+    def __init__(
+        self,
+        name: str,
+        default: Optional[
+            Union[
+                str,
+                float,
+                int,
+                bool,
+                Dict[str, Any],
+                Callable[[], Union[str, float, int, bool, Dict[str, Any]]],
+            ]
+        ] = None,
+        type: Optional[
+            Union[Type[str], Type[float], Type[int], Type[bool], JSONTypeClass]
+        ] = None,
+        help: Optional[str] = None,
+        required: bool = False,
+        show_default: bool = True,
```

Comment:
this changes the semantics a bit since e.g. `required` is `False` if not defined, not `None` as before. I hope it doesn't change the behavior in `click` but needs extra attention

Reply:
Looking at the page here: https://click.palletsprojects.com/en/7.x/api/#click.Parameter, required defaults to False when not specified. For type, we do default to str (so made it that explicitly), for show_default, we had a line that set it to True if None. This seems to make it clearer to the user (hopefully).